### DISCUSSION TITLE
[typescript-urql-svelte] New Codegen Plugin Request

### DISCUSSION THREAD

**Discussion by trevyn** [dotansimha/graphql-code-generator#5431] [Ideas] [answered]
The existing `typescript-urql` plugin is specific to React, and [Svelte](https://svelte.dev/) is awesome. I have [`@urql/svelte`](https://formidable.com/open-source/urql/docs/api/svelte/) up and running, and would love to see a `typescript-urql-svelte` plugin for it.

As of today, the Svelte [repo](https://github.com/sveltejs/svelte) has 41.4k stars, so it should be around a while. :)

**Comment by ardatan** [comment]
@trevyn That sounds great! Would you start with a draft PR to help us? We don't know much about Svelte.

**Comment by trevyn** [comment]
Sure, I'll take a look. For now, [here's my quick-and-dirty `@urql/svelte` impl](https://github.com/trevyn/turbo/blob/375e7a42fb707970dc462ce25272e19968c64e8b/frontend/App.svelte). I honestly am not sure about the TypeScript part yet, but I'll keep this thread updated and submit a draft PR if I figure out a sensible plugin. :)

**Comment by trevyn** [comment]
Ok, I did a reasonable draft of what the codegen output could look like, and verified it works:

[`App.svelte`](https://github.com/trevyn/turbo/blob/45fe80fa5cd05fd9d381aedc44e2588fbcdad623/frontend/App.svelte) (usage)
[`graphql_ops.svelte`](https://github.com/trevyn/turbo/blob/45fe80fa5cd05fd9d381aedc44e2588fbcdad623/frontend/graphql_ops.svelte) (would be codegen'd)

This is just in straight JS, I'll need to add TypeScript.

I also looked at the plugin code for `typescript-urql` to understand how to actually make the plugin. It looks OK, so I'll probably do that over the next few days. (Unless this is already enough, and you have someone that wants to take it on?)

**Comment by dotansimha** [comment]
Moving to Discussions :)

**Comment by kitten** [comment, marked as answer]
Wouldn't it be more interesting here to promote `TypedDocumentNode` as the "preferred" solution across the board? It's supported in more cases (like methods on the `Client`), it is framework-agnostic, and it supports inferred types that are as good as with generated utilities (or in the case of React/Preact, generated wrapper hooks)

**Reply by dotansimha** [reply]
I'm totally agree with that, and we are still aiming to promote `TypedDocumentNode` as the ideal usage :)

**Reply by trevyn** [reply]
Oh man, I *totally* missed that.

For anyone else wanting to use TypeScript + urql + Svelte + codegen, just use the [`typed-document-node` plugin](https://graphql-code-generator.com/docs/plugins/typed-document-node), and you can pass that straight to `@urql/svelte`:

```svelte
<script lang="ts">
 import * as gql from "./your-codegenned-file";  // .ts
 import { operationStore, query } from "@urql/svelte";

 const listCards = query(operationStore(gql.ListCardsDocument));
</script>
```

See also: https://the-guild.dev/blog/typed-document-node
